File: benchmarks/scala-stm/scala-stm-library/lib/sb7-java-v1.2/src/stmbench7/impl/core/AtomicPartImpl.java
package stmbench7.impl.core;

import stmbench7.backend.ImmutableCollection;
import stmbench7.core.AtomicPart;
import stmbench7.core.CompositePart;
import stmbench7.core.Connection;
import stmbench7.impl.backend.SmallSetImpl;

/**
 * STMBench7 benchmark Atomic Part (see the specification).
 * Default implementation.
 */
public class AtomicPartImpl extends DesignObjImpl implements AtomicPart {

    private int x, y;
    protected SmallSetImpl<Connection> to, from;
    private CompositePart partOf;

    public AtomicPartImpl(int id, String type, int buildDate, int x, int y) {
    	super(id, type, buildDate);
    	this.x = x;
    	this.y = y;
    	to = new SmallSetImpl<Connection>();
    	from = new SmallSetImpl<Connection>();
    }

    public AtomicPartImpl(AtomicPartImpl source) {
    	super(source);
    	this.x = source.x;
    	this.y = source.y;
    	this.to = new SmallSetImpl<Connection>(source.to);
    	this.from = new SmallSetImpl<Connection>(source.from);
    	this.partOf = source.partOf;
    }
    
    public void connectTo(AtomicPart destination, String type, int length) {
    	Connection connection = new ConnectionImpl(this, destination, type, length);
    	to.add(connection);
    	destination.addConnectionFromOtherPart(connection.getReversed());
    }

    public void addConnectionFromOtherPart(Connection connection) {
    	from.add(connection);
    }
    
    public void setCompositePart(CompositePart partOf) {
    	this.partOf = partOf;
    }

    public int getNumToConnections() {
    	return to.size();
    }

    public ImmutableCollection<Connection> getToConnections() {
    	return to.immutableView();
    }

    public ImmutableCollection<Connection> getFromConnections() {
    	return from.immutableView();
    }
    
    public CompositePart getPartOf() {
    	return partOf;
    }

    public void swapXY() {
    	int tmp = y;
    	y = x;
    	x = tmp;
    }
    
    public int getX() {
    	return x;
    }
    
    public int getY() {
    	return y;
    }
    
    public void clearPointers() {
    	x = y = 0;
    	to = null;
    	from = null;
    	partOf = null;
    }

	public int compareTo(AtomicPart part) {
		return id - part.getId();
	}
	
	@Override
	public boolean equals(Object obj) {
		if(! (obj instanceof AtomicPart)) return false;
		return super.equals(obj);
	}

	@SuppressWarnings("unchecked")
	@Override
	public AtomicPartImpl clone() {
		AtomicPartImpl clone = (AtomicPartImpl) super.clone();
		clone.from = (SmallSetImpl<Connection>) from.clone();
		clone.to = (SmallSetImpl<Connection>) to.clone();
		return clone;
	}
	
	@Override
	public String toString() {
		return super.toString() + ", x=" + x + ", y=" + y + ", partOf=[" + partOf + "]" +
			", to={ " + to.size() + " connections }, from={ " + from.size() + " connections }"; 
	}
}

File: benchmarks/scala-stm/src/main/scala/stmbench7/scalastm/AtomicPartImpl.scala
/* scala-stm - (c) 2009-2011, Stanford University, PPL */

package stmbench7.scalastm

import scala.concurrent.stm.atomic
import scala.concurrent.stm.Ref
import scala.concurrent.stm.TSet

import stmbench7.core.AtomicPart
import stmbench7.core.CompositePart
import stmbench7.core.Connection
import stmbench7.impl.core.ConnectionImpl

class AtomicPartImpl(id0: Int, typ0: String, bd0: Int, x0: Int, y0: Int)
  extends DesignObjImpl(id0, typ0, bd0)
  with AtomicPart {

  private val x = Ref(x0)
  private val y = Ref(y0)
  private val partOf = Ref(null: CompositePart).single
  private val from = TSet.empty[Connection].single // this is the equivalent of SmallSetImpl
  private val to = TSet.empty[Connection].single

  override def connectTo(dest: AtomicPart, typ: String, length: Int): Unit = {
    val c = new ConnectionImpl(this, dest, typ, length)
    to += c
    dest.addConnectionFromOtherPart(c.getReversed)
  }

  override def addConnectionFromOtherPart(c: Connection): Unit = { from += c }
  override def setCompositePart(po: CompositePart): Unit = { partOf() = po }
  override def getNumToConnections: Int = to.size
  override def getToConnections = new ImmutableSetImpl[Connection](to)
  override def getFromConnections = new ImmutableSetImpl[Connection](from)
  override def getPartOf: CompositePart = partOf()

  override def swapXY(): Unit = {
    atomic { implicit t =>
      y() = x.swap(y())
    }
  }
  override def getX: Int = x.single()
  override def getY: Int = y.single()

  override def clearPointers(): Unit = {
    atomic { implicit t =>
      x() = 0
      y() = 0
      to.clear()
      from.clear()
      partOf() = null
    }
  }

  // Comparable[AtomicPart]
  override def compareTo(rhs: AtomicPart): Int =
    getId - rhs.getId // subtraction is faithful to reference impl
}
